### PR TITLE
Typing compatible Channels

### PR DESCRIPTION
**Pull request by mosquito**
Full typing compatibility for `Channel`

image

### CODE REVIEW

**Comment by mosquito** [comment]
@tbug Do you want me to add py.typed and CI Actions for type checking etc.

**Review by dizballanze — DISMISSED** [review]

**Review by Alviner — DISMISSED** [review]

File: aiochannel/channel.py [hunk 1]
```diff
@@ -177,21 +188,21 @@ def close(self):
...
+    async def __anext__(self) -> T:
         try:
             data = await self.get()
         except ChannelClosed:
             raise StopAsyncIteration
         else:
             return data
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[T]:
```
Comment (leenr) [inline aiochannel/channel.py]:
```suggestion
    def __iter__(self) -> Iterator[T]:
```

File: aiochannel/channel.py [hunk 2]
```diff
@@ -1,20 +1,31 @@
 from .errors import ChannelClosed, ChannelFull, ChannelEmpty
 from collections import deque
-from asyncio import Event, Future, get_event_loop
+from asyncio import Event, Future, get_event_loop, AbstractEventLoop
+from typing import Any, Generic, Iterable, TypeVar, Deque
```
Comment (leenr) [inline aiochannel/channel.py]:
```suggestion
from asyncio import AbstractEventLoop, Event, Future, get_event_loop
from typing import Any, Deque, Generic, Iterable, TypeVar
```

File: aiochannel/channel.py [hunk 3]
```diff
@@ -38,42 +49,42 @@ def __init__(self, maxsize=0, *, loop=None):
     def _init(self):
```
Comment (leenr) [inline aiochannel/channel.py]:
```suggestion
    def _init(self) -> None:
```

**Review by leenr — CHANGES_REQUESTED** [review]

**Comment by tudborg** [comment]
Would be great @mosquito 👍

**Comment by mosquito** [comment]
@tbug How do you look at modernizing this project a little, adding pyproject.toml and transferring all the project settings there. I think that it will be cleaner and easier than maintaining self-written scripts. 

<summmary>For example, now, a lot of superfluous gets into the sdist:
<details>

```
    $ python3.10 setup.py sdist
    running sdist
    running egg_info
    writing aiochannel.egg-info/PKG-INFO
    writing dependency_links to aiochannel.egg-info/dependency_links.txt
    writing top-level names to aiochannel.egg-info/top_level.txt
    reading manifest file 'aiochannel.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching 'README.md'
    writing manifest file 'aiochannel.egg-info/SOURCES.txt'
    running check
    creating aiochannel-1.1.1
    creating aiochannel-1.1.1/aiochannel
    creating aiochannel-1.1.1/aiochannel.egg-info
    creating aiochannel-1.1.1/tests
    creating aiochannel-1.1.1/tests/examples
    creating aiochannel-1.1.1/tests/regressions
    creating aiochannel-1.1.1/tools
    copying files to aiochannel-1.1.1...
    copying MANIFEST.in -> aiochannel-1.1.1
    copying README.rst -> aiochannel-1.1.1
    copying setup.cfg -> aiochannel-1.1.1
    copying setup.py -> aiochannel-1.1.1
--> copying tox.ini -> aiochannel-1.1.1**
    copying aiochannel/__init__.py -> aiochannel-1.1.1/aiochannel
    copying aiochannel/channel.py -> aiochannel-1.1.1/aiochannel
    copying aiochannel/errors.py -> aiochannel-1.1.1/aiochannel
    copying aiochannel/version.py -> aiochannel-1.1.1/aiochannel
    copying aiochannel.egg-info/PKG-INFO -> aiochannel-1.1.1/aiochannel.egg-info
    copying aiochannel.egg-info/SOURCES.txt -> aiochannel-1.1.1/aiochannel.egg-info
    copying aiochannel.egg-info/dependency_links.txt -> aiochannel-1.1.1/aiochannel.egg-info
    copying aiochannel.egg-info/not-zip-safe -> aiochannel-1.1.1/aiochannel.egg-info
    copying aiochannel.egg-info/top_level.txt -> aiochannel-1.1.1/aiochannel.egg-info
--> copying tests/__init__.py -> aiochannel-1.1.1/tests
--> copying tests/test_channel.py -> aiochannel-1.1.1/tests
--> copying tests/test_channel_35features.py -> aiochannel-1.1.1/tests
--> copying tests/examples/__init__.py -> aiochannel-1.1.1/tests/examples
--> copying tests/examples/consume_channel_python35.py -> aiochannel-1.1.1/tests/examples
--> copying tests/examples/simple_usage.py -> aiochannel-1.1.1/tests/examples
--> copying tests/regressions/__init__.py -> aiochannel-1.1.1/tests/regressions
--> copying tests/regressions/test_invalid_state_error_during_close.py -> aiochannel-1.1.1/tests/regressions
--> copying tools/clean.sh -> aiochannel-1.1.1/tools
--> copying tools/test-requirements.txt -> aiochannel-1.1.1/tools
    Writing aiochannel-1.1.1/setup.cfg
    Creating tar archive
    removing 'aiochannel-1.1.1' (and everything under it)
```

</details>
</summary>

**Comment by mosquito** [comment]
@tbug a [modern project structure](https://github.com/mosquito/aiochannel/pull/1) awaiting your review

**Comment by tudborg** [comment]
@mosquito neat. Will take a look ASAP.

I haven't done python professionally for years, so I'm a little rusty 🙃

**Comment by tudborg** [comment]
Looks good @mosquito.

I've updated tox.ini so it should not be runnable with the current workers.

Please squash your commits down into 1 commit, then I'll merge it in.

**Comment by mosquito** [comment]
@tbug it's already here.

**Comment by mosquito** [comment]
@tbug I would ask you not to hurry with the release right away, maybe the other of my ideas will be included here.

**Comment by tudborg** [comment]
Sure thing @mosquito. We'll take a look at your other changes as well 👍

**Comment by tudborg** [comment]
`v1.2.0` released on PyPI

**Pull request merged**